### DISCUSSION TITLE
When should I load the collection?

### DISCUSSION THREAD

**Discussion by hf185** [milvus-io/milvus#14522] [Q&A and General discussion]
在什么时机去Load Collections是最好的呢?
我尝试过在搜索前是load，搜索完成后release，这样反复操作后，容易出现一些奇怪的现象。比如：Collection里面明明有entity，但是搜索结果为空。或者某次load的时间特别长。
如果是提前load的话，那应该在什么时候release呢？还是说可以让它一直在内存里？但是在内存里我又遇到一种情况就是内存暴涨，当时的QPS是在60左右，直接把服务器256G内存全部占用完了，然后Milvus又连接不上了。
这个问题让我很困惑，load和release的时机该怎么掌控?
----------------
When is the best time to go to Load Collections?
I tried to load before the search and release after the search is complete. After repeated operations, some strange phenomena are prone to appear. For example: There is an entity in the Collection, but the search result is empty. Or a certain load takes a very long time.
If it is loaded in advance, when should it be released? Or is it possible to keep it in memory all the time? But in the memory, I encountered another situation where the memory skyrocketed. At that time, the QPS was around 60, and the server 256G memory was used up, and then Milvus could not be connected.
This question confuses me, how should I control the timing of load and release?

**Comment by yhmo** [comment]
The 2.0 doesn't allow users to set cache size. Once you call load_collection(), it always try to load the whole collection data into CPU memory. That means user need to make sure the machine's memory is larger than the collection data, otherwise, it will OOM.
Once you finish searching collection A, you want to search collection B. You need to release A and load B. Basically, just keep the data that is required to search in memory, release other unnecessary data.

**Reply by hf185** [reply]
> 

If I only have one collection, but 100W data is added every day, is it suitable to stay in memory all the time?

**Reply by yhmo** [reply]
The data size is unlimited? Then your search action is targeted to the whole dataset, or just search in a part of data in recent days?

**Reply by hf185** [reply]
1. The vector is 256 bits, and there are two Int64 and one Int32 Fields.
2. Different scenarios need to search for data in different time periods. There is a timestamp in the Field to filter the data

**Reply by hadim** [reply]
In case of a single very large collection, is that possible to load it on multiple nodes? (split per partition or segments)

**Reply by yhmo** [reply]
> In case of a single very large collection, is that possible to load it on multiple nodes? (split per partition or segments)

The 2.0 has an API load_balance(), user can specify to balance segments from one query node to other query nodes.
```
def load_balance(src_node_id, dst_node_ids=None, sealed_segment_ids=None, timeout=None, using="default"):
    """
    Do load balancing operation from source query node to destination query node.

    :param src_node_id: The source query node id to balance.
    :type  src_node_id: int

    :param dst_node_ids: The destination query node ids to balance.
    :type  dst_node_ids: list[int]

    :param sealed_segment_ids: Sealed segment ids to balance.
    :type  sealed_segment_ids: list[int]
```

**Reply by yhmo** [reply]
> 1. The vector is 256 bits, and there are two Int64 and one Int32 Fields.
> 2. Different scenarios need to search for data in different time periods. There is a timestamp in the Field to filter the data

In 2.0, the search() API has a parameter named "travel_timestamp", user can specify a timestamp to search data from the timestamp to latest:
```
 def search(self, data, anns_field, param, limit, expr=None, partition_names=None,
               output_fields=None, timeout=None, round_decimal=-1, **kwargs):
        """
        :param kwargs:
            * *travel_timestamp* (``int``) --
              Users can specify a timestamp in a search to get results based on a data view
                        at a specified point in time.
```
But to unload the old data, you need to split data into partitions, and call release_partition() to release old data.
For example, you create one partition for each day, so you have 5 partition for 5 days, their names are "d_1", "d_2", "d_3", "d_4", "d_5". You only want to search the latest 3 days data("d_3", "d_4", "d_5"), then you can call release_partitions(["d_1", "d_2"]). In the sixth day, you got "d_6", now you want to search in "d_4", "d_5", "d_6", then you can call release_partitions(["d_3"]).

**Reply by hf185** [reply]
> In 2.0, the search() API has a parameter named "travel_timestamp", user can specify a timestamp to search data from the timestamp to latest:

ok, thank you. 

Is it appropriate to keep this collection in memory for a long time?
In addition, can you specify that the similarity of the response data is above X when searching?

**Reply by yhmo** [reply]
> 

Loading a huge dataset from storage is very time-consuming, especially with storage based on network. So it is better to keep it in memory if you constantly do searching on this collection.

"specify that the similarity of the response data is above X when searching", this function is not yet implemented, we call it "search by range". Maybe available in v2.1.

**Reply by hf185** [reply]
> > 
> 
> Loading a huge dataset from storage is very time-consuming, especially with storage based on network. So it is better to keep it in memory if you constantly do searching on this collection.
> 
> "specify that the similarity of the response data is above X when searching", this function is not yet implemented, we call it "search by range". Maybe available in v2.1.

I understand, thank you.
"search by range" I think it's very important, and I look forward to it!
